NuRadioMC/EvtGen/generator.py

```diff
@@ -934,123 +823,116 @@ def generate_surface_muons(filename, n_events, Emin, Emax,
-    data_sets["n_interaction"] = np.ones(n_events, dtype=np.int)
-    data_sets["vertex_times"] = np.zeros(n_events, dtype=np.float)
-
-    # generate neutrino flavors randomly
-
-    data_sets["flavors"] = np.array([flavor[i] for i in np.random.randint(0, high=len(flavor), size=n_events)])
-
-    # generate energies randomly
-    if(spectrum == 'log_uniform'):
-        data_sets["energies"] = 10 ** np.random.uniform(np.log10(Emin), np.log10(Emax), n_events)
-    elif(spectrum.startswith("E-")):  # enerate an E^gamma spectrum.
-        gamma = float(spectrum[1:])
-        gamma += 1
-        Nmin = (Emin) ** gamma
-        Nmax = (Emax) ** gamma
-
-        def get_inverse_spectrum(N, gamma):
-            return np.exp(np.log(N) / gamma)
-
-        data_sets["energies"] = get_inverse_spectrum(np.random.uniform(Nmax, Nmin, size=n_events), gamma)
-    else:
-        logger.error("spectrum {} not implemented".format(spectrum))
-        raise NotImplementedError("spectrum {} not implemented".format(spectrum))
-
-    # generate charged/neutral current randomly
-    data_sets["interaction_type"] = [ '' ] * n_events
-
-    # generate inelasticity
-    data_sets["inelasticity"] = np.zeros(n_events)
+    data_sets_fiducial = {}
+    proposal_time = 0
 
-    data_sets["energies"] = np.array(data_sets["energies"])
-    data_sets["muon_energies"] = np.copy(data_sets["energies"])
+    set_volume_attributes(volume, proposal=False, attributes=attributes)
+    n_events = attributes['n_events']  # important! the number of events might have been increased by the set_volume_attributes function
+    n_batches = int(np.ceil(n_events / max_n_events_batch))
+    for i_batch in range(n_batches):  # do generation of events in batches
+        data_sets = {}
+        n_events_batch = max_n_events_batch
+        if(i_batch + 1 == n_batches):  # last batch?
+            n_events_batch = n_events - (i_batch * max_n_events_batch)
+        data_sets["xx"], data_sets["yy"], data_sets["zz"] = generate_vertex_positions(attributes=attributes, n_events=n_events_batch)
+        data_sets["zz"] = np.zeros_like(data_sets["yy"])  # muons interact at the surface
 
-    data_sets_fiducial = {}
+        # generate neutrino vertices randomly
+        data_sets["azimuths"] = np.random.uniform(phimin, phimax, n_events_batch)
+        # zenith directions are distruted as sin(theta) (to make the distribution istotropic) * cos(theta) (to account for the projection onto the surface)
+        data_sets["zeniths"] = np.arcsin(np.random.uniform(np.sin(thetamin) ** 2, np.sin(thetamax) ** 2, n_events_batch) ** 0.5)
 
-    import time
-    init_time = time.time()
-    # Initialising data_sets_fiducial with empty values
-    for key, value in data_sets.items():
-        data_sets_fiducial[key] = []
+        data_sets["event_group_ids"] = np.arange(i_batch * max_n_events_batch, i_batch * max_n_events_batch + n_events_batch) + start_event_id
```

Comment:
Won't this generate floats instead of integers?

Reply:
good catch. I changed it